dcos_installer/upgrade.py

```diff
@@ -0,0 +1,128 @@
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+set -o errexit -o nounset -o pipefail
+
+source /opt/mesosphere/environment.export
+
+# Check if this is a terminal, and if colors are supported, set some basic
+# colors for outputs
+if [ -t 1 ]; then
+    colors_supported=$(tput colors)
+    if [[ $colors_supported -ge 8 ]]; then
+    RED='\e[1;31m'
+    BOLD='\e[1m'
+    NORMAL='\e[0m'
+    fi
+fi
+
+if [[ $EUID -ne 0 ]]; then
+    echo "This script must be run as root" 1>&2
+    exit 1
+fi
+
+
+# check for version of dc/os upgrading from
+version=`grep "version" /opt/mesosphere/etc/dcos-version.json | cut -d '"' -f 4`
+if [ $version != {{ dcos_version }} ]; then
+   echo "Not the correct version, please rerun upgrade script with the installed dcos version ${version}"
+   exit 1
+fi
+
+echo "Fetching packages"
+
+pkgpanda fetch --repository-url={{ bootstrap_url }} {{ cluster_packages }}
+
+echo "Activating packages"
+
+pkgpanda activate --no-block {{ cluster_packages }}
+
+# check if we are on a master node
+if [ -f /etc/mesosphere/roles/master ]; then
+   # run exhibitor migration script here
+   until dcos-shell dcos-exhibitor-migrate-perform
+   do
+        status=$?
+        case $status in
+             1) echo "Waiting for exhibitor endpoint"
+                sleep 10
+                ;;
+             2) echo "Could not read from exhibitor"
+                echo "Contact Support"
+                exit
+                ;;
+             4) echo "Rolling update in progress"
+                exit
+                ;;
+             8) echo "At least one config value does not have the expected pre migration value,
+                      \ and automatic migration can not take place"
+                echo "Contact Support"
+                exit
+                ;;
+             16) echo "Attempting to start the rolling update failed due to a non 200 response from exhibitor"
+                 echo "Contact Support"
+                 exit
+                 ;;
+        esac
+   done
+fi
+
+echo "Node upgrade complete"
+
+"""
+
+
+def generate_node_upgrade_script(gen_out, current_version, serve_dir=SERVE_DIR):
+
+    bootstrap_url = gen_out.arguments['bootstrap_url']
+
+    package_list = ' '.join(package['id'] for package in gen_out.cluster_packages.values())
+
+    bash_script = gen.template.parse_str(node_upgrade_template).render({
+        'dcos_image_commit': util.dcos_image_commit,
+        'generation_date': util.template_generation_date,
+        'bootstrap_url': bootstrap_url,
+        'cluster_packages': package_list,
+        'dcos_version': current_version})
+
+    hash_string = hash_str(bash_script)
```

Comment:
It's hard to tell, but I think this means that successive re-runs of this command with the same version against the same `genconf/` directory will output the upgrade script to the same location, regardless of any config changes, since script content doesn't depend on config. Probably not what we want -- the purpose of the random directory is to prevent successive runs from overwriting each other. The only requirement for this directory in the design doc is that it be "randomly named", so I don't think we need to name it based on its content like this. A truly random string like a `uuid.uuid4().hex` would fit the bill here.